crates/flashblocks-rpc/src/cache.rs

```diff
@@ -0,0 +1,279 @@
+    receipts_cache: HashMap<TxHash, OpTransactionReceipt>,
+}
+
+impl FlashblocksCacheInner {
+    pub fn new(chain_spec: Arc<OpChainSpec>) -> Self {
+        Self {
+            chain_spec,
+            builder: FlashblockBuilder::new(),
+            block: None,
+            balance_cache: HashMap::new(),
+            nonce_cache: HashMap::new(),
+            receipts_cache: HashMap::new(),
+        }
+    }
+
+    pub fn get_block(&self, full: bool) -> Option<RpcBlock<Optimism>> {
+        let block = match &self.block {
+            Some(block) => block,
+            None => return None,
+        };
+
+        let header: alloy_consensus::Header = block.header.clone();
+        let transactions = block.body.transactions.to_vec();
+
+        if full {
+            let transactions_with_senders = transactions
+                .into_iter()
+                .zip(block.body.recover_signers().unwrap());
+            let converted_txs = transactions_with_senders
+                .enumerate()
+                .map(|(idx, (tx, sender))| {
+                    let signed_tx_ec_recovered = Recovered::new_unchecked(tx.clone(), sender);
+                    let tx_info = TransactionInfo {
+                        hash: Some(tx.tx_hash()),
+                        block_hash: Some(block.header.hash_slow()),
+                        block_number: Some(block.number),
+                        index: Some(idx as u64),
+                        base_fee: block.base_fee_per_gas,
+                    };
+                    transform_tx(signed_tx_ec_recovered, tx_info)
+                })
+                .collect();
+            Some(RpcBlock::<Optimism> {
+                header: Header::from_consensus(header.seal_slow(), None, None),
+                transactions: BlockTransactions::Full(converted_txs),
+                uncles: Vec::new(),
+                withdrawals: None,
+            })
+        } else {
+            let tx_hashes = transactions.into_iter().map(|tx| tx.tx_hash()).collect();
+            Some(RpcBlock::<Optimism> {
+                header: Header::from_consensus(header.seal_slow(), None, None),
+                transactions: BlockTransactions::Hashes(tx_hashes),
+                uncles: Vec::new(),
+                withdrawals: None,
+            })
+        }
+    }
+
+    pub fn reset(&mut self) {
+        self.block = None;
+        self.builder = FlashblockBuilder::new();
+        self.balance_cache.clear();
+        self.nonce_cache.clear();
+        self.receipts_cache.clear();
+    }
+
+    pub fn process_payload(&mut self, payload: FlashblocksPayloadV1) -> eyre::Result<()> {
+        // Convert metadata with error handling
+        let metadata: Metadata = match serde_json::from_value(payload.metadata.clone()) {
+            Ok(m) => m,
+            Err(e) => {
+                return Err(eyre::eyre!("Failed to deserialize metadata: {}", e));
+            }
+        };
+
+        if payload.index == 0 {
+            self.reset();
```

Comment:
there is a time gap between the node getting the latest block and the first flashblock (triggered by block building instead of fcu without attributes). wondering if the cache should be reset on a new block notification instead